javascript/src/main/pages/AnalyzeMessageData/HistogramOfMessagesByUser.js

```diff
@@ -1,9 +1,81 @@
-import React from "react";
+import React, { useState } from "react";
+import { fetchWithToken } from "../../utils/fetch";
+import { useAuth0 } from "@auth0/auth0-react";
+import BootstrapTable from 'react-bootstrap-table-next';
+import { Row, Col, Button, Form } from "react-bootstrap";
+
+const columns = [{
+    dataField: 'week',
+    text: 'Week'
+},
+{
+    dataField: 'count',
+    text: 'Message Count'
+}
+];
+
+const msInDay = 86400*1000;
 
 const HistogramOfMessagesByUser = () => {
-    return <div>
-        <h1>Histogram Of Messages By User</h1>
-    </div>
+    const { getAccessTokenSilently: getToken } = useAuth0();
+    const [selectedUser, setSelectedUser] = useState('');
+    const [inputName, setInputName] = useState('');
+    const [perWeekHistogramData, setPerWeekHistogramData] = useState([]);
+
+    const handleSelectUserOnSubmit = async () => { 
+        setSelectedUser(inputName);
+        const url = `/api/members/messages/usersearch?searchUser=${inputName}`;
+        const options = {
+            method: 'GET',
+        }
+        const messages = await fetchWithToken(url, getToken, options);
+
+        let histogramData = {};
+
+        messages.forEach(i => {
+            //mulitply by 1000 to get timestamp in ms
+            const date = new Date(i.ts*1000);
+
+            const weekStartDate = new Date(date.getTime() - msInDay*date.getDay());
+            const weekEndDate = new Date(date.getTime() + msInDay*(6-date.getDay()));
+            const weekString = `${weekStartDate.toLocaleDateString()} - ${weekEndDate.toLocaleDateString()}`;
+
+            histogramData[weekString] = histogramData[weekString] > 0? histogramData[weekString]+1 : 1;
+        })
+
+        let tableData = [];
+        Object.keys(histogramData).forEach(week => tableData.push({
+            week: week,
+            count: histogramData[week]
+        }));
```

Comment:
I realize now that instead of using a `forEach` loop here, you can instead use `map`. They both iterate through each element in an array, but `map` takes the return value of the function and creates a new array with the return values, which is what we're doing here.

If you're up to the challenge, do you want to change it?

Reply:
Changes look good otherwise. Let me know how you want to proceed.

Reply:
I suggest that we merge this, and offer a 5 point story for refactoring with a map.

Reply:
That sounds good to me